### PR TITLE
add support for function as match

### PR DESCRIPTION
**Pull request by yocontra**
adds the ability to use an arbitrary function as the match, this lets you use more complex logic than just matching error strings, like:

```js
// retry on all errors but 404, currently not possible with existing impl
match: (err) => err.status !== 404
```

### CODE REVIEW

**Comment by dhensby** [comment]
This would be quite useful :)

**Comment by yocontra** [comment]
Rebased - @mickhansen anything you need before considering this?

File: index.js [hunk 1]
```diff
@@ -38,11 +38,11 @@ module.exports = function retryAsPromised(callback, options) {
     name: options.name || callback.name || 'unknown'
   };
 
-  var report = options.report;
-
-  // Massage match option into array so we can blindly treat it as such later
-  if (!Array.isArray(options.match)) options.match = [options.match];
+  if (typeof options.match !== 'function' && !Array.isArray(options.match)) {
```
Comment (mickhansen) [inline index.js]: Support `[function]` just for conviences sake.

Comment (yocontra) [reply index.js]: @mickhansen Makes sense, will do

**Comment by yocontra** [comment]
@mickhansen Updated. I also went and cleaned up/optimized some of the code that does the matching + added support for booleans as well.

There is an unrelated changed I added that I hope is uncontroversial - tacking the original error onto the timeout error, so when it gets reported to newrelic/sentry/etc. that information is not lost. If you are opposed to that change I can back it out.

File: index.js [hunk 2]
```diff
@@ -4,15 +4,31 @@ var Promise = require('any-promise');
 var util = require('util');
 var format = util.format;
 
-function TimeoutError(message) {
+function TimeoutError(message, err) {
   Error.call(this);
   Error.captureStackTrace(this, TimeoutError);
   this.name = 'TimeoutError';
   this.message = message;
+  this.original = err;
```
Comment (mickhansen) [inline index.js]: previous might be more accurate

Comment (yocontra) [reply index.js]: @mickhansen Was going for something similar to http://docs.sequelizejs.com/class/lib/errors/index.js~DatabaseError.html but I can change it if you want.

Comment (mickhansen) [reply index.js]: I figured @contra and it makes sense when it's a wrapped error, in this case it's not a wrapped error but and independent error where errors might have occured in the past.

Comment (yocontra) [reply index.js]: @mickhansen Updated

**Comment by mickhansen** [comment]
`v3.2.0`

**Comment by yocontra** [comment]
@mickhansen Thanks!

**Pull request merged**